design/ceph-csi-operator.md

```diff
@@ -0,0 +1,89 @@
+
+## Upgraded clusters
+
+For upgrade clusters, we'll fetch the CSI settings and cluster connections details from configmaps like `rook-ceph-operator-config` and `rook-ceph-csi-config`. After fetching the CSI settings and cluster connections
+details we can fill the new Ceph-CSI operator cr `CephCSICephConfig` with connection details and other CR's. Once we have the settings we can set `USE_CSI_OPERATOR:` is to `true` it will also disable the CSI `ROOK_CSI_DISABLE_DRIVER`.
+
+**Note:** Not much changes are anticipated for external mode clusters.
+
+## New Ceph-CSI operator CRDs
+
+### Ceph-CSI-Operator configuration
+
+The CSI operator brings new CRDS and API for users to configure settings. These CRDs are:
+
+### CephCSICephCluster CRD
+
+[This CRD](https://github.com/ceph/ceph-csi-operator/blob/main/docs/design/operator.md#cephCSIcephcluster-crd) stores connection and configuration details for a single Ceph cluster and provide the information to be used by multiple CSI drivers.
+
+Any setting in this CR either needs to be generated by Rook, or wrapped by CephCluster CR under `spec.csi` or `rook-ceph-cluster` helm chart `values.yaml`.
+
+### CephCSIConfig CRD
+
+[This CRD](https://github.com/ceph/ceph-CSI-operator/blob/main/docs/design/operator.md#cephCSIconfig-crd) contains details about CephFS, RBD, and NFS configuration to be used when communicating with Ceph. Include a reference to a CephCSICephCluster holding the connection information for the target Ceph cluster.
+
+The Rook operator would generate an instance of these CR for each of CephFilesystemSubVolumeGroup and CephBlockPoolRadosNamespace CRs.
+
+### CephCSIOperatorConfig CRD
+
+[This CRD](https://github.com/ceph/ceph-CSI-operator/blob/main/docs/design/operator.md#cephCSIoperatorconfig-crd)  manages operator-level configurations and offers a place to overwrite settings for CSI drivers. This CRD is a namespace-scoped CRD and a single CR named ceph-CSI-operator-config can be created by an user inside the operator's namespace.
+
+### CephCSIDriver CRD
+
+Rook will define three instances of [this CRD](https://github.com/ceph/ceph-CSI-operator/blob/main/docs/design/operator.md#cephCSIdriver-crd), one each for RBD, CephFS, and NFS for managing the installation, lifecycle management, and configuration for CephFS, RBD, and NFS CSI drivers.
+The CRs will be empty (all the settings are defined as defaults in the previous CR), or else some small set of default driver-specific settings.
+
+### CephCSIConfigMapping
+
+[This CRD](https://github.com/ceph/ceph-CSI-operator/blob/main/docs/design/operator.md#cephCSIconfigmapping) contains a mapping between local and remote Ceph cluster configurations.It also provides a mapping between the CephCSIConfigurations.
+Rook will define these mappings when the rbd mirroring is enabled.
+
+## Open Questions
+
+1. Should automate the deployment of CSI-operator internally or leave it to users to deploy via manifests.
```

Comment:
Do we still have this open question, can we settle on this one.
cc @travisn

Reply:
I think we did settle it above, so this question can be resolved. For experimental mode it will be deployed with separate manifests, and when stable we will automate it.

Reply:
@travisn which means the user need to create the operatorConfig and driver CR's for all 3 drivers?

Reply:
I would summarize the following:
1. During a new install, the operatorConfig and driver CRs will be created by default either with rook's operator.yaml, or the helm chart. If the user wants to modify those CRs, they fully own them. Rook will not reconcile them.
2. During an upgrade from the old csi driver, Rook will automatically create the operatorConfig and driver CRs according to the settings in the old csi driver. But then Rook will no longer reconcile them. This will be an upgrade feature after the csi operator is stable, and not while experimental.

Reply:
@travisn @Madhu-1 are we on the same page to ahead with PR?

Reply:
LGTM, perhaps you can just remove this open question since it's answered now.
```suggestion
```

Reply:
done, removed

Reply:
@travisn if Rook won't reconcile it we have a problem with the downstream day 2 operation which needs discussion

Reply:
ocs-operator can always deploy what Rook does not take on.